### PR TITLE
feature: cross-check penalty details with E5

### PR DESCRIPTION
**Pull request by akaJSaunders**
### JIRA link

https://companieshouse.atlassian.net/browse/LFA-1264

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                await this.chSdk(new OAuth2(accessToken!)).lateFilingPenalties.getPenalties(companyNumber);
+
+            if (penalties.httpStatusCode !== OK) {
+                throw new Error(`AppealDetailActionProcessor: failed to get penalties from pay API with status code ${penalties.httpStatusCode} with access token ${accessToken} and base url ${API_URL}`);
+            }
+
+            const items = penalties.resource!.items
+                .filter(penalty =>
+                    penalty.type === 'penalty' &&
+                    penalty.id === penaltyReference
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:63]: You're going to have to strip down the alphabetic character for new PR numbers. See tech notes i added.
Also, doing this check will only work for new penalty reference numbers. We should check if the penalty reference is in the old format and if it is we should avoid checking the penalty id.

Comment (franciscog-kainos) [reply src/controllers/validators/PenaltyDetailsValidator.ts:63]: And yes, for old PR numbers we potentially will have multiple ones of type late filling

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        try {
+
+            const penalties: Resource<PenaltyList> =
+                await this.chSdk(new OAuth2(accessToken!)).lateFilingPenalties.getPenalties(companyNumber);
+
+            if (penalties.httpStatusCode !== OK) {
+                throw new Error(`AppealDetailActionProcessor: failed to get penalties from pay API with status code ${penalties.httpStatusCode} with access token ${accessToken} and base url ${API_URL}`);
+            }
+
+            let items: Penalty[];
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:60]: Initialise items to [] so you don't have to check if it's undefined later

Comment (franciscog-kainos) [reply src/controllers/validators/PenaltyDetailsValidator.ts:60]: you could shorten this to:
```ts
let items: Penalty[] = penalties.resource.items.filter(penalty => penalty.type === 'penalty');
if (modernPenaltyReferenceRegex.test(penaltyReference)){
    items = items.filter(penalty.id === penaltyReference.substring(1));
}
```

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    .filter(penalty => penalty.type === 'penalty');
+            }
+
+            if (!items || items.length === 0) {
+                loggerInstance().error(`No penalties for ${companyNumber} match the reference number ${penaltyReference}`);
+                return this.createValidationResultWithErrors();
+            }
+
+        } catch (err) {
+            if (err.message === 'Cannot read property \'map\' of null' || 'Cannot read property \'etag\' of null') {
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:82]: It's a good idea to make constants for these

File: src/controllers/validators/PenaltyDetailsValidator.ts (hunk 2 above)
Comment (akaJSaunders) [reply src/controllers/validators/PenaltyDetailsValidator.ts:60]: Good one

File: test/controllers/PenaltyDetailsController.test.ts [hunk 4]
```diff
@@ -70,6 +98,122 @@ describe('PenaltyDetailsController', () => {
...
+            const appeal = {
+                penaltyIdentifier: {
+                    companyNumber: 'SC123123',
+                    penaltyReference: 'A12312323',
+                }
+            } as Appeal;
+
+            const lfpService = createSubstituteOf<LateFilingPenaltyService>();
+
+            lfpService.getPenalties('SC123123').throws(new Error('Cannot read property \'map\' of null'));
```
Comment (franciscog-kainos) [inline test/controllers/PenaltyDetailsController.test.ts:112]: Change the error text for the constant you created please

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                return this.createValidationResultWithErrors();
+            }
+
+        } catch (err) {
+            if (err.message === mapErrorMessage || etagErrorMessage) {
+                loggerInstance().error(`company number ${companyNumber} could not be found: ${err}`);
+                return this.createValidationResultWithErrors();
+            }
+
+            throw new Error(`Can't access API: ${err}`);
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:86]:
```suggestion
            throw new Error(`${AppealDetailActionProcessor.name}: Can't access API: ${err}`);
```

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        const mapErrorMessage = 'Cannot read property \'map\' of null';
+        const etagErrorMessage =  'Cannot read property \'etag\' of null';
+
+        try {
+
+            const penalties: Resource<PenaltyList> =
+                await this.chSdk(new OAuth2(accessToken!)).lateFilingPenalties.getPenalties(companyNumber);
+
+            if (penalties.httpStatusCode !== OK) {
+                throw new Error(`AppealDetailActionProcessor: failed to get penalties from pay API with status code ${penalties.httpStatusCode} with access token ${accessToken} and base url ${API_URL}`);
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:64]:
```suggestion
                throw new Error(`${AppealDetailActionProcessor.name}: failed to get penalties from pay API with status code ${penalties.httpStatusCode} with access token ${accessToken} and base url ${API_URL}`);
```

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            }
+
+            if (!items || items.length === 0) {
+                loggerInstance().error(`No penalties for ${companyNumber} match the reference number ${penaltyReference}`);
+                return this.createValidationResultWithErrors();
+            }
+
+        } catch (err) {
+            if (err.message === mapErrorMessage || etagErrorMessage) {
+                loggerInstance().error(`company number ${companyNumber} could not be found: ${err}`);
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:82]:
```suggestion
                loggerInstance().error(`${AppealDetailActionProcessor.name}: company number ${companyNumber} could not be found: ${err}`);
```

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            const modernPenaltyReferenceRegex: RegExp = /^[A-Z][0-9]{8}$/;
+
+            let items: Penalty[] = penalties.resource!.items.filter(penalty => penalty.type === 'penalty');
+
+            if (modernPenaltyReferenceRegex.test(penaltyReference)){
+                items = items.filter(penalty => penalty.id === penaltyReference.substring(1));
+            }
+
+            if (!items || items.length === 0) {
+                loggerInstance().error(`No penalties for ${companyNumber} match the reference number ${penaltyReference}`);
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:76]:
```suggestion
                loggerInstance().error(`${AppealDetailActionProcessor.name}: No penalties for ${companyNumber} match the reference number ${penaltyReference}`);
```

File: src/controllers/validators/PenaltyDetailsValidator.ts (hunk 6 above)
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:64]: My bad

```suggestion
                throw new Error(`PenaltyDetailsValidator: failed to get penalties from pay API with status code ${penalties.httpStatusCode} with access token ${accessToken} and base url ${API_URL}`);
```

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            const modernPenaltyReferenceRegex: RegExp = /^[A-Z][0-9]{8}$/;
+
+            let items: Penalty[] = penalties.resource!.items.filter(penalty => penalty.type === 'penalty');
+
+            if (modernPenaltyReferenceRegex.test(penaltyReference)){
+                items = items.filter(penalty => penalty.id === penaltyReference.substring(1));
+            }
+
+            if (!items || items.length === 0) {
+                loggerInstance().error(`${AppealDetailActionProcessor.name}: No penalties for ${companyNumber} match the reference number ${penaltyReference}`);
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:76]:
```suggestion
                loggerInstance().error(`${PenaltyDetailsValidator.name}: No penalties for ${companyNumber} match the reference number ${penaltyReference}`);
```

File: src/controllers/validators/PenaltyDetailsValidator.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                return this.createValidationResultWithErrors();
+            }
+
+        } catch (err) {
+            if (err.message === mapErrorMessage || etagErrorMessage) {
+                loggerInstance().error(`${AppealDetailActionProcessor.name}: company number ${companyNumber} could not be found: ${err}`);
```
Comment (franciscog-kainos) [inline src/controllers/validators/PenaltyDetailsValidator.ts:82]:
```suggestion
                loggerInstance().error(`${PenaltyDetailsValidator.name}: company number ${companyNumber} could not be found: ${err}`);
```
